### PR TITLE
Category axis doesn't consider offscreen labels when requesting space

### PR DESCRIPTION
**Pull request by hellochar**
Category Axis will now only render/consider the ticks that will actually be drawn. This fixes a bug where pan/zoomed category axes might have a lot of unnecessary whitespace.

### CODE REVIEW

**Comment by adidahiya** [comment]
screencaps / gifs? is this smooth now?

File: src/axes/categoryAxis.ts [hunk 1]
```diff
@@ -134,10 +134,10 @@ export class Category extends Axis<string> {
    * @param {Scales.Category} scale - The scale being downsampled. Defaults to this Axis' scale.
    * @return {DownsampleInfo} an object holding the resultant domain and new stepWidth.
    */
-  public getDownsampleInfo(scale: Scales.Category = <Scales.Category> this._scale): DownsampleInfo {
+  public getDownsampleInfo(scale: Scales.Category = <Scales.Category> this._scale, domain = scale.invertRange()): DownsampleInfo {
```
Comment (themadcreator) [inline src/axes/categoryAxis.ts:142]: I didn't know you could reference previous args in later default args. Neat

File: src/axes/categoryAxis.ts [hunk 2]
```diff
@@ -370,7 +367,6 @@ export class Category extends Axis<string> {
     // hide ticks and labels that overflow the axis
     this._showAllTickMarks();
     this._showAllTickLabels();
-    this._hideOverflowingTickLabels();
```
Comment (themadcreator) [inline src/axes/categoryAxis.ts:378]: Why don't we need this anymore?

File: src/scales/categoryScale.ts [hunk 3]
```diff
@@ -50,6 +50,19 @@ export class Category extends Scale<string, number> implements TransformableScal
     this._outerPadding = Category._convertToPlottableOuterPadding(0.5, d3InnerPadding);
   }
 
+  /**
+   * Return a clone of this category scale without any included values providers.
```
Comment (themadcreator) [inline src/scales/categoryScale.ts]: I'm glad you're encapsulating this into a method. Can you include in this comment some info about what **IS** included in the partial "clone" of this scale.

**Review by themadcreator — CHANGES_REQUESTED** [review]
Just a comment update request

**Comment by hellochar** [comment]
There's some jank for how this behaves when pan/zoom and downsampling are enabled:

Zoom suddenly stops because the axis grows:


Layout jankily jumps around because the specific labels being drawn (and therefor the height) change while panning:


Thoughts?

**Comment by helenkg** [comment]
@hellochar I would assume once we do vertical labels and truncate them at fixed height, the height of individual labels won't change when zooming and panning?

**Comment by adidahiya** [comment]
> once we do vertical labels and truncate them at fixed height

@helenkg this still won't fix the problem because some labels might be at the full height & truncated while others are not close to the truncation limit

@hellochar in general I'm not a fan of the behavior shown in the above gifs; what if you waited until pan/zoom completed before redrawing the layout?

**Comment by hellochar** [comment]
@adidahiya not a fan of tying re-compute to panzoom interactions since that adds a lot of complexity. Let me try debouncing requestedSpace() and see how that feels

**Comment by adidahiya** [comment]
@hellochar can you elaborate on the code complexity just a bit? I think it could work out really well actually from a user's perspective.

Also curious if @pkwi has any thoughts

**Comment by hellochar** [comment]
throttling doesn't work. Also, with maxWidth and 90 degree angles, it still has some jank:



@pkwi thoughts here?

**Comment by hellochar** [comment]
@adidahiya here's "wait until pan is finished to recalculate":

ezgif com-video-to-gif 9

Definitely feels better. The zoom issue is still there but I think I'm okay with leaving that in until people complain.

**Comment by themadcreator** [comment]
Okay, i think i finally get how this works. I was just missing the fact that your pass in the value of `.invertRange()` as the domain. Looks good, and this should work really well with sheared ticks

**Review by themadcreator — APPROVED** [review]

**Comment by themadcreator** [comment]
I'm gonna merge this because I want to pull these changes in to test sheared ticks. If there are still more changes to address the "jank", that can be a followup PR

**Pull request merged**